lib/galaxy/model/base.py

```diff
@@ -41,6 +42,25 @@
 REQUEST_ID = ContextVar("request_id", default=_request_state.copy())
 
 
+def commit(session: Union[scoped_session, Session, "SessionlessContext"]):
+    """
+    Wraps SQLAlchemy's Session.commit method. Should be called from galaxy instead of using
+    SQLAlchemy's method directly.
+    """
+    err = None
+    try:
+        session.commit()
+    except PendingRollbackError as e:
+        err = e
```

Comment:
So now we haven't committed and the calling context has no idea that we didn't commit. That doesn't make sense to me.

Reply:
My reasoning was this. A `PendingRollbackError` may be raised due to the session trying to access the database (a) after a server disconnect which invalidates the outer, logical (not database) transaction; or (b) after a database transaction rollback due to failed commit, which results in the current logical transaction being in `inactive` state. 

In most cases, if a commit causes a database transaction rollback, SA will raise an error that we don't handle, which is not bad - at least we'll see the error message. But I'm sure there are places in our codebase where we have a block of code wrapped in a try/except that handles a generic `Exception` without rolling back the logical transaction, after which, as soon as we try to access the database, we'll get the `PendingRollbackError`. Here's one such example: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/authnz/custos_authnz.py#L361 . I'm sure it's not the only one.

By calling `commit(session)`, we at least ensure that the logical transaction is rolled back, and when the session tries to access the database after that, a new transaction begins. Of course this does not solve the issue of the silenced error raised during the commit - but that's a separate problem (that also should be eventually addressed), but it ensures that the previous error will not affect subsequent commits.

It's a partial solution only - it handles commits, but not database access in general. If a disconnect happens, or a bad commit is silenced, and then the session tries to access the database in any other way (explicit read via select or implicit read when accessing a lazy loaded attribute), we'll still have the same problem. However, given that we commit left and right (after which a new logical transaction is started as soon as the session needs one), I thought this still could be helpful as a partial solution. If you think it's not a good approach, I'm OK with closing this.